public/app/plugins/panel/gauge/Thresholds.tsx

```diff
@@ -0,0 +1,306 @@
+        <div className={rowStyle} key={`${threshold.index}-${index}`}>
+          <div className="threshold-row-inner">
+            <div className="threshold-row-color">
+              {threshold.color && (
+                <div className="threshold-row-color-inner">
+                  <ColorPicker
+                    color={threshold.color}
+                    onChange={color => this.onChangeThresholdColor(threshold, color)}
+                  />
+                </div>
+              )}
+            </div>
+            <input
+              className="threshold-row-input"
+              type="text"
+              onChange={event => this.onChangeThresholdValue(event, threshold)}
+              value={threshold.value}
+              onBlur={this.onBlur}
+            />
+            {threshold.canRemove ? (
+              <div onClick={() => this.onRemoveThreshold(threshold)} className="threshold-row-remove">
+                <i className="fa fa-times" />
+              </div>
+            ) : (
+              <div className="threshold-row-label">{threshold.label}</div>
+            )}
+          </div>
+        </div>
+      );
+    });
+  }
+
+  insertAtIndex(index) {
+    const { thresholds } = this.state;
+
+    // If thresholds.length is greater or equal to 3
+    // it means a user has added one threshold
+    if (thresholds.length < 3 || index < 0) {
+      return 1;
+    }
+
+    return index;
+  }
+
+  renderIndicatorSection(index) {
+    const { thresholds } = this.state;
+    const indicators = thresholds.length - 1;
+
+    if (index === 0 || index === thresholds.length) {
+      return (
+        <div
+          key={index}
+          className="indicator-section"
+          style={{
+            height: `calc(100%/${indicators})`,
+          }}
+        >
+          <div
+            onClick={() => this.onAddThreshold(this.insertAtIndex(index - 1))}
+            style={{
+              height: '100%',
+              background: this.getIndicatorColor(index),
+            }}
+          />
+        </div>
+      );
+    }
+
+    return (
+      <div
+        key={index}
+        className="indicator-section"
+        style={{
+          height: `calc(100%/${indicators})`,
+        }}
+      >
+        <div
+          onClick={() => this.onAddThreshold(this.insertAtIndex(index))}
+          style={{
+            height: '50%',
+            background: this.getIndicatorColor(index),
+          }}
+        />
+        <div
+          onClick={() => this.onAddThreshold(this.insertAtIndex(index + 1))}
+          style={{
+            height: `50%`,
+            background: this.getIndicatorColor(index),
+          }}
+        />
+      </div>
+    );
+  }
+
+  renderIndicator() {
+    const { thresholds } = this.state;
+
+    const indicators = thresholds.length - 1;
+
+    const sections = [];
+
+    for (let i = 0; i < indicators; i++) {
+      sections.push(this.renderIndicatorSection(i));
+    }
+
+    return sections;
+  }
```

Comment:
I'm not sure if you think we loose readability with this:
```
renderIndicator() {
  return this.state.thresholds.map((threshold, i) => this.renderIndicatorSection(i));
}
```

Reply:
did a version of that :)